File: src/main/java/ThirtyDayLeetCodingChallenge/Day6_GroupAnagrams.java
package ThirtyDayLeetCodingChallenge;

import java.util.*;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

public class Day6_GroupAnagrams {
    public static void main(String[] args) {
        System.out.println(groupAnagrams(new String[]{"eat", "tea", "tan", "ate", "nat", "bat"}));
    }

    public static List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> dict = new HashMap();
        for (String str : strs) {
            String sorted = Arrays.toString(str.chars().mapToObj(c -> (char) c).sorted().toArray());
            if (!dict.containsKey(sorted)) dict.put(sorted, new ArrayList());
            dict.get(sorted).add(str);
        }

        return dict.values().stream().collect(Collectors.toList());
    }
}

File: src/main/kotlin/ThirtyDayLeetCodingChallenge/Day6_GroupAnagrams.kt
fun main() {
    println(groupAnagrams(arrayOf("eat", "tea", "tan", "ate", "nat", "bat")))
}

fun groupAnagrams(strs: Array<String>): List<List<String>> {
    val dict = mutableMapOf<String, ArrayList<String>>()
    for (str in strs) {
        val sorted = str.toCharArray().sorted().joinToString("")
        if (!dict.containsKey(sorted)) dict[sorted] = ArrayList()
        dict[sorted]!!.add(str)
    }

    return dict.values.toList()
}
